### ISSUE TITLE
Support monochrome BMP ?

### ISSUE THREAD

**Issue by lufengd3** [image-rs/image#1646]
I would like to be able create monochrome bmp,  1 bit per pixel.

My specific use case for this functionality is e-ink screen displays. The basic device only support monochrome bitmap.

**Comment by fintelia** [comment]
Does your use case involve encoding, decoding, or both?

**Comment by lufengd3** [comment]
> Does your use case involve encoding, decoding, or both?

Currently only use encoding.

**Comment by Sheepyhead** [comment]
I'll second the need for 1bpp bmp, in this case to write the bmp to an eink screen. I'd like to be able to create the bmp in code, then write it to a file, and there doesn't seem to be a crate that supports this for 1bpp bmps

**Comment by fintelia** [comment]
At this point, the groundwork is in place if anyone wants to work on a PR: The `write_image` method can be called with `ColorType::L1` (currently it returns an error) and [this comment](https://github.com/image-rs/image/issues/2425#issuecomment-2692454046) describes the expected data layout.

**Comment by ajacquet-evolis** [comment]
Supporting the 1bpp BMP encoding/decoding seems to "only" require to edit `BmpDecoder` and `BmpEncoder` in order to accept the `ExtendedColorType::L1`.

But you would still need to provide the image data (`&[u8]`) correctly based on your previous link, which means it's up to the user to make the conversion from `L8` to 1bpp for example, right ?
Would it make sense to integrate 1bpp image in the `ColorType` and `DynamicImage` enum to ease the conversion and export of such image type ? This would for example allow to use the `DynamicImage::save` function.

**Comment by fintelia** [comment]
> Would it make sense to integrate 1bpp image in the `ColorType and `DynamicImage` enum to ease the conversion and export of such image type ? This would for example allow to use the `DynamicImage::save` function.

No. The problem is that a bunch of places (both in this library and in user code) assume that pixels take an integer number of bytes to store. The most obvious is [`ColorType::bytes_per_pixel`](https://docs.rs/image/latest/image/enum.ColorType.html#method.bytes_per_pixel) but there's a bunch of others that show up because all Rust structs have to be an integer number of bytes

**Comment by ajacquet-evolis** [comment]
Ok, thanks for the insight.

What about introducing a new utilitary function somewhere (on `DynamicImage` maybe ?) to export the image as a 1bpp `Vec<u8>` that would be a valid input for `BmpEncode::WriteImage` with `ExtendedColorType::L1` ?
This means that the output `Vec<u8>` of this new function would fill the following requirement from your comment:
> sub 8-bit pixel data is packed with multiple samples in a single byte. Rows of pixels are padded to byte boundaries.

**Comment by fintelia** [comment]
I think it would be good to have a function like that. Honestly, the hardest part is likely the API design: a single conversion function isn't too bad but there's dozens of image layouts and theoretically people could want to pairwise convert between any of them. How do we decided which ones to support? What would a consistent API for that look like? etc.

The function itself should pretty much just be (untested, but a rough sketch):

```rust
fn l8_to_l1(img: &[u8], width: u32, height: u32, threshold: u8) -> Vec<u8> {
    let mut output = vec![0; width.div_ceil(8) as usize * height as usize];
    for y in 0..height {
        for x in 0..width {
            if img[i] >= threshold {
                output[y * width.div_ceil(8) + x / 8] |= 1 << (x % 8);
            }
        }
    }
    output
}

**Comment by RunDevelopment** [comment]
Resolved by #2953
